### ISSUE TITLE
clippy::question_mark shouldn't emit when return value isn't used.

### ISSUE THREAD

**Issue by kangalioo** [rust-lang/rust-clippy#7967]
Lint name: `clippy::question_mark`

I tried this code:

```rust
pub struct Message;
pub struct MessageUpdateEvent { content: Option<String> }

pub fn process_message_update(
    user_msg_update: MessageUpdateEvent,
) -> Option<Message> {
    // [some code]
    
    // If message content wasn't touched, don't re-run command
    if user_msg_update.content.is_none() {
        return None;
    }
    
    // [more code]
    
    Some(Message)
}
```
(for more context, see https://github.com/kangalioo/poise/blob/19bcc0afef2c77533a4e2f0a4aaffa4baea74f2a/src/prefix/track_edits.rs#L85-L92)

I expected to see this happen: no warning is emitted

Why: semantically, the `?` operator is meant for propagating exceptional values which the current function doesn't want to deal with. The semantics in this code are different, so usage of `?` would be confusing and make it harder to figure what is actually happening here and why

Instead, this happened: clippy suggested to replace the check with `user_msg_update.content.as_ref()?;`

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.55.0 (c8dfcfe04 2021-09-06)
binary: rustc
commit-hash: c8dfcfe046a7680554bf4eb612bad840e7631c4b
commit-date: 2021-09-06
host: x86_64-unknown-linux-gnu
release: 1.55.0
LLVM version: 12.0.1
```

**Comment by dswij** [comment]
`question_mark` actually checks for these kind of cases, `if foo.is_none() { return None; }` [from the description here](https://github.com/rust-lang/rust-clippy/blob/0d283cc44c1ac9e0b61ccc66eb1544abed075dfb/clippy_lints/src/question_mark.rs#L44-L58).

Although I don't think it's accurate to say it's FP, I can see why one would prefer explicitly writing these out.
